File: lang/python/bench_time.py
#!/usr/bin/env python3
"""Minimal timing shim: outputs parse=X.XXX stream=X.XXX (median ms, medium fixture)."""
import sys, os, time
sys.path.insert(0, os.path.dirname(__file__))
import cxlib

with open(os.path.join(os.path.dirname(__file__), '..', '..', 'fixtures', 'bench', 'bench_medium.cx')) as f:
    medium = f.read()

def time_median(fn, n=100, warmup=20):
    for _ in range(warmup): fn()
    ts = []
    for _ in range(n):
        t0 = time.perf_counter()
        fn()
        ts.append((time.perf_counter() - t0) * 1000)
    ts.sort()
    return ts[n // 2]

parse_med  = time_median(lambda: cxlib.parse(medium))
stream_med = time_median(lambda: list(cxlib.stream(medium)))
print(f'parse={parse_med:.3f} stream={stream_med:.3f}')

File: lang/rust/cxlib/src/bin/bench_time.rs
use std::time::Instant;

fn main() {
    let medium = std::fs::read_to_string("../../../fixtures/bench/bench_medium.cx")
        .expect("run from lang/rust/cxlib/");

    let parse_med  = time_median(100, 20, || { cxlib::parse(&medium).unwrap(); });
    let stream_med = time_median(100, 20, || { cxlib::stream(&medium).unwrap(); });
    println!("parse={:.3} stream={:.3}", parse_med, stream_med);
}

fn time_median(n: usize, warmup: usize, mut f: impl FnMut()) -> f64 {
    for _ in 0..warmup { f(); }
    let mut times: Vec<f64> = (0..n).map(|_| {
        let t0 = Instant::now();
        f();
        t0.elapsed().as_secs_f64() * 1000.0
    }).collect();
    times.sort_by(|a, b| a.partial_cmp(b).unwrap());
    times[n / 2]
}
